src/decoder/builder.rs

```diff
@@ -0,0 +1,307 @@
+}
+
+impl<R> Default for DecoderBuilder<R> {
+    fn default() -> Self {
+        Self {
+            data: None,
+            settings: Settings::default(),
+        }
+    }
+}
+
+impl<R: Read + Seek + Send + Sync + 'static> DecoderBuilder<R> {
+    /// Creates a new decoder builder with default settings.
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use std::fs::File;
+    /// use rodio::decoder::DecoderBuilder;
+    ///
+    /// let file = File::open("audio.mp3").unwrap();
+    /// let decoder = DecoderBuilder::new()
+    ///     .with_data(file)
+    ///     .build()
+    ///     .unwrap();
+    /// ```
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Sets the input data source to decode.
+    pub fn with_data(mut self, data: R) -> Self {
+        self.data = Some(data);
+        self
+    }
+
+    /// Sets the byte length of the stream.
+    /// This is required for:
+    /// - Reliable seeking operations
+    /// - Duration calculations in formats that lack timing information (e.g. MP3, Vorbis)
+    ///
+    /// The byte length should typically be obtained from file metadata:
+    /// ```no_run
+    /// use std::fs::File;
+    /// use rodio::Decoder;
+    ///
+    /// let file = File::open("audio.mp3").unwrap();
+    /// let len = file.metadata().unwrap().len();
+    /// let decoder = Decoder::builder()
+    ///     .with_data(file)
+    ///     .with_byte_len(len)
+    ///     .build()
+    ///     .unwrap();
+    /// ```
+    ///
+    /// Alternatively, it can be obtained by seeking to the end of the stream.
+    pub fn with_byte_len(mut self, byte_len: u64) -> Self {
+        self.settings.byte_len = Some(byte_len);
+        self
+    }
+
+    /// Enables or disables coarse seeking.
+    ///
+    /// This needs byte_len to be set. Coarse seeking is faster but less accurate:
+    /// it may seek to a position slightly before or after the requested one,
+    /// especially when the bitrate is variable.
+    pub fn with_coarse_seek(mut self, coarse_seek: bool) -> Self {
+        self.settings.coarse_seek = coarse_seek;
+        self
+    }
+
+    /// Enables or disables gapless playback.
+    ///
+    /// When enabled, removes silence between tracks for formats that support it.
+    pub fn with_gapless(mut self, gapless: bool) -> Self {
+        self.settings.gapless = gapless;
+        self
+    }
+
+    /// Sets a format hint for the decoder.
+    ///
+    /// When known, this can help the decoder to select the correct codec.
+    /// Common values are "mp3", "wav", "flac", "ogg", etc.
+    pub fn with_hint(mut self, hint: &str) -> Self {
+        self.settings.hint = Some(hint.to_string());
+        self
+    }
+
+    /// Sets a mime type hint for the decoder.
+    ///
+    /// When known, this can help the decoder to select the correct demuxer.
+    /// Common values are "audio/mpeg", "audio/vnd.wav", "audio/flac", "audio/ogg", etc.
+    pub fn with_mime_type(mut self, mime_type: &str) -> Self {
```

Comment:
why is there both a mime type and a hint? (To be clear, this is just me not understanding). Its all audio right?

Reply:
Can't speak for Symphonia but I would explain as follows:
- for some sources (like HTTP streams) you may know the MIME content type but not the extension
- for other sources (like files) it may be the other way around

Further: as per #612 the `.m4x` extensions are quite meaningless, but at least the MIME type is correct. Symphonia [understand both](https://github.com/pdeljanov/Symphonia/blob/ef9bbd8dd147b05cc911dafe0ae3663ae81b692d/symphonia-format-isomp4/src/demuxer.rs#L313) but setting the MIME type seems a tad bit more accurate here to me.

Reply:
and the hints speed up the choice of the correct decoder? Or is it possible to pick a wrong format when doing so automatically (like `build_impl` does)?

Reply:
Yes, it speeds up probing by prioritizing the hinted demuxer/decoder. The probe continue with other demuxers/decoders if the provided hints did not yield the expected result.

Reply:
In that case maybe we could add the word faster to the docs, concretely: `this can help the decoder to select the correct demuxer faster`. Since now a user might think the mime/hint is needed for correct functionality in some edge case.